File: problems/数学与位运算/338.比特位计数.py
#!/usr/bin/env python3
"""
338. 比特位计数
"""
def count_bits(n: int) -> list[int]:
    dp = [0] * (n + 1)
    for i in range(1, n + 1):
        dp[i] = dp[i >> 1] + (i & 1)
    return dp

if __name__ == "__main__":
    # 测试用例
    print(count_bits(2))   # 预期: [0, 1, 1]
    print(count_bits(5))   # 预期: [0, 1, 1, 2, 1, 2]
    print(count_bits(0))   # 预期: [0]
    print(count_bits(1))   # 预期: [0, 1]

File: problems/数学与位运算/338.比特位计数.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> countBits(int n) {
    vector<int> dp(n+1, 0);
    for (int i = 1; i <= n; i++)
        dp[i] = dp[i>>1] + (i & 1);
    return dp;
}

int main() {
    // 测试用例
    auto r1 = countBits(2);
    for (int x : r1) cout << x << " "; cout << endl;  // 预期: 0 1 1
    auto r2 = countBits(5);
    for (int x : r2) cout << x << " "; cout << endl;  // 预期: 0 1 1 2 1 2
    auto r3 = countBits(0);
    for (int x : r3) cout << x << " "; cout << endl;  // 预期: 0
    auto r4 = countBits(1);
    for (int x : r4) cout << x << " "; cout << endl;  // 预期: 0 1
    return 0;
}
